File: example/canport/canport_set_zero.cpp
#include "canport.h"

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdio>
#include <thread>

static std::atomic<bool> exit_flag(false);


int main()
{
    constexpr auto kControlPeriod = std::chrono::microseconds(1000);
    std::signal(SIGINT, [](int) { exit_flag.store(true); });

    CanPort can_port(1, {1, 2, 3});

    auto next_tick = std::chrono::steady_clock::now();
    while (!exit_flag.load())
    {
        for (const auto& item : can_port.map_motors_state)
        {
            const float id = item.first;
            can_port.pos_vel_acc(id, 0.0f, 0.314f, 3.14f);
        }
        can_port.send();

        for (const auto& item : can_port.map_motors_state)
        {
            const float id = item.first;
            const motor_state_t& motor = item.second;
            std::printf("ID: %2d, mode: %2d, fluat: %2d, pos: %2.3f, vel: %2.3f, tor: %2.3f\n",
                        static_cast<int>(id), motor.mode, motor.fault,
                        motor.position, motor.velocity, motor.torque);
        }

        next_tick += kControlPeriod;
        std::this_thread::sleep_until(next_tick);
    }

    can_port.stop();
    can_port.send();
    return 0;
}

File: example/python/canport/canport_set_zero.py
import time

from hightorque_robot import CanPort


def main():
    kControlPeriod = 0.001
    can_port = CanPort(1, [1])

    can_port.motor_zero_pos_reset()

    next_tick = time.monotonic()
    while True:
        can_port.request_motor_state()
        for item in can_port.map_motors_state.items():
            id = item[0]
            motor = item[1]
            print(
                "ID: {:2d}, mode: {:2d}, fault: {:2d}, pos: {:2.3f}, "
                "vel: {:2.3f}, tor: {:2.3f}".format(
                    int(id),
                    motor.mode,
                    motor.fault,
                    motor.position,
                    motor.velocity,
                    motor.torque,
                )
            )

        next_tick += kControlPeriod
        time.sleep(max(0.0, next_tick - time.monotonic()))

    can_port.stop()
    can_port.send()


if __name__ == "__main__":
    main()
